File: daybreak/src/daybreak/adapters/terminal/windows_terminal.py
import logging
import os
from pathlib import Path

from daybreak.config import config

from .jsonc import dump_json_file, load_jsonc_file

logger = logging.getLogger("daybreak")


class WindowsTerminalAdapter:
    name = "windows_terminal"

    def apply_mode(self, mode: str, theme_name: str, palette: dict = None):
        if os.name != "nt":
            return

        target_scheme = config.get(
            "integrations",
            f"windows_terminal_{mode}_scheme",
            "One Half Light" if mode == "light" else "One Half Dark",
        )

        for settings_path in _iter_windows_terminal_settings():
            self._apply_scheme(settings_path, target_scheme)

    def _apply_scheme(self, path: Path, target_scheme: str):
        try:
            data = load_jsonc_file(path)
        except Exception as exc:
            logger.warning(f"Windows Terminal: Failed to parse {path}: {exc}")
            return

        changed = False
        profiles = data.setdefault("profiles", {})
        defaults = profiles.setdefault("defaults", {})
        if defaults.get("colorScheme") != target_scheme:
            defaults["colorScheme"] = target_scheme
            changed = True

        profile_list = profiles.get("list", [])
        if isinstance(profile_list, list):
            for entry in profile_list:
                if isinstance(entry, dict) and "colorScheme" in entry:
                    if entry.get("colorScheme") != target_scheme:
                        entry["colorScheme"] = target_scheme
                        changed = True

        if changed:
            try:
                dump_json_file(path, data)
                logger.info(f"Windows Terminal: Applied colorScheme '{target_scheme}' to {path}")
            except Exception as exc:
                logger.warning(f"Windows Terminal: Failed writing {path}: {exc}")


def _iter_windows_terminal_settings():
    local_app_data = os.environ.get("LOCALAPPDATA")
    if not local_app_data:
        return []

    base = Path(local_app_data)
    results = []
    seen = set()

    packaged_root = base / "Packages"
    if packaged_root.exists():
        for package_dir in packaged_root.iterdir():
            if not package_dir.is_dir():
                continue
            if not package_dir.name.startswith("Microsoft.WindowsTerminal"):
                continue
            candidate = package_dir / "LocalState" / "settings.json"
            if candidate.exists() and str(candidate) not in seen:
                results.append(candidate)
                seen.add(str(candidate))

    unpackaged_candidate = base / "Microsoft" / "Windows Terminal" / "settings.json"
    if unpackaged_candidate.exists() and str(unpackaged_candidate) not in seen:
        results.append(unpackaged_candidate)

    return results

File: src/adapters/terminal/windowsTerminal.ts
import * as fs from 'fs';
import * as path from 'path';
import { config } from '../../config';
import { Palette } from '../../colors';
import { loadJsoncFile, dumpJsonFile } from './jsonc';

const logger = {
  info: (m: string) => {},
  warn: (m: string) => console.warn('[daybreak]', m),
};

export class WindowsTerminalAdapter {
  name = 'windows_terminal';

  applyMode(mode: string, _themeName: string, _palette?: Palette): void {
    if (process.platform !== 'win32') return;

    const targetScheme = config.get(
      'integrations',
      `windows_terminal_${mode}_scheme`,
      mode === 'light' ? 'One Half Light' : 'One Half Dark',
    );

    for (const settingsPath of iterWindowsTerminalSettings()) {
      this._applyScheme(settingsPath, targetScheme);
    }
  }

  private _applyScheme(filePath: string, targetScheme: string): void {
    let data: any;
    try {
      data = loadJsoncFile(filePath);
    } catch (exc) {
      logger.warn(`Windows Terminal: Failed to parse ${filePath}: ${exc}`);
      return;
    }

    let changed = false;
    if (!data.profiles) data.profiles = {};
    if (!data.profiles.defaults) data.profiles.defaults = {};
    if (data.profiles.defaults.colorScheme !== targetScheme) {
      data.profiles.defaults.colorScheme = targetScheme;
      changed = true;
    }

    const profileList = data.profiles.list;
    if (Array.isArray(profileList)) {
      for (const entry of profileList) {
        if (entry && typeof entry === 'object' && 'colorScheme' in entry) {
          if (entry.colorScheme !== targetScheme) {
            entry.colorScheme = targetScheme;
            changed = true;
          }
        }
      }
    }

    if (changed) {
      try {
        dumpJsonFile(filePath, data);
        logger.info(`Windows Terminal: Applied colorScheme '${targetScheme}' to ${filePath}`);
      } catch (exc) {
        logger.warn(`Windows Terminal: Failed writing ${filePath}: ${exc}`);
      }
    }
  }
}

function iterWindowsTerminalSettings(): string[] {
  const localAppData = process.env['LOCALAPPDATA'];
  if (!localAppData) return [];

  const results: string[] = [];
  const seen = new Set<string>();

  const packagedRoot = path.join(localAppData, 'Packages');
  if (fs.existsSync(packagedRoot)) {
    try {
      for (const entry of fs.readdirSync(packagedRoot)) {
        if (!entry.startsWith('Microsoft.WindowsTerminal')) continue;
        const candidate = path.join(packagedRoot, entry, 'LocalState', 'settings.json');
        if (fs.existsSync(candidate) && !seen.has(candidate)) {
          results.push(candidate);
          seen.add(candidate);
        }
      }
    } catch { /* ignore */ }
  }

  const unpackaged = path.join(localAppData, 'Microsoft', 'Windows Terminal', 'settings.json');
  if (fs.existsSync(unpackaged) && !seen.has(unpackaged)) {
    results.push(unpackaged);
  }

  return results;
}
